File: Medium/649. Dota2 Senate/Solution.java
class Solution {
    public String predictPartyVictory(String senate) {
        int n = senate.length();
        Queue<Integer> R = new LinkedList<>(), D = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            if(senate.charAt(i) == 'R') R.add(i);
            else D.add(i);
        }

        while(!R.isEmpty() && !D.isEmpty()) {
            int r = R.poll(), d = D.poll();
            if(r < d) R.add(r + n);
            else D.add(d + n);
        }

        return (R.isEmpty() ? "Dire" : "Radiant");
    }
}

File: Medium/649. Dota2 Senate/Solution.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        queue<int> R, D;
        int n = (int) senate.size();

        for(int i = 0; i < n; i++) {
            if(senate[i] == 'R') R.emplace(i);
            else D.emplace(i);
        }

        while(!D.empty() and !R.empty()) {
            int r = R.front(), d = D.front();
            R.pop(), D.pop();
            if(r < d) R.emplace(r + n);
            else D.emplace(d + n);
        }

        return R.empty() ? "Dire" : "Radiant";
    }
};